src/Spanner/Database.php

```diff
@@ -0,0 +1,1441 @@
+     *
+     * @param callable $operation The operations to run in the transaction.
+     *        **Signature:** `function (Transaction $transaction)`.
+     * @param array $options [optional] {
+     *     Configuration Options
+     *
+     *     @type int $maxRetries The number of times to attempt to apply the
+     *           operation before failing. **Defaults to ** `10`.
+     *     @type bool $singleUse If true, a Transaction ID will not be allocated
+     *           up front. Instead, the transaction will be considered
+     *           "single-use", and may be used for only a single operation. Note
+     *           that in a single-use transaction, only a single operation may
+     *           be executed, and rollback is not available. **Defaults to**
+     *           `false`.
+     * }
+     * @return mixed The return value of `$operation`.
+     * @throws \RuntimeException
+     */
+    public function runTransaction(callable $operation, array $options = [])
+    {
+        $options += [
+            'maxRetries' => self::MAX_RETRIES,
+        ];
+
+        // There isn't anything configurable here.
+        $options['transactionOptions'] = $this->configureTransactionOptions();
+
+        $session = $this->selectSession(SessionPoolInterface::CONTEXT_READWRITE);
+
+        $attempt = 0;
+        $startTransactionFn = function ($session, $options) use (&$attempt) {
+            if ($attempt > 0) {
+                $options['isRetry'] = true;
+            }
+
+            $transaction = $this->operation->transaction($session, $options);
+
+            $attempt++;
+            return $transaction;
+        };
+
+        $delayFn = function (\Exception $e) {
+            if (!($e instanceof AbortedException)) {
+                throw $e;
+            }
+
+            $delay = $e->getRetryDelay();
+            time_nanosleep($delay['seconds'], $delay['nanos']);
+        };
+
+        $transactionFn = function ($operation, $session, $options) use ($startTransactionFn) {
+            $transaction = call_user_func_array($startTransactionFn, [
+                $session,
+                $options
+            ]);
+
+            $res = call_user_func($operation, $transaction);
+
+            $active = $transaction->state() === Transaction::STATE_ACTIVE;
+            $singleUse = $transaction->type() === Transaction::TYPE_SINGLE_USE;
+            if ($active && !$singleUse) {
+                $transaction->rollback($options);
+                throw new \RuntimeException('Transactions must be rolled back or committed.');
+            }
+
+            return $res;
+        };
+
+        $retry = new Retry($options['maxRetries'], $delayFn);
+
+        try {
+            return $retry->execute($transactionFn, [$operation, $session, $options]);
+        } finally {
+            $session->setExpiration();
+        }
+    }
+
+    /**
+     * Insert a row.
+     *
+     * Mutations are committed in a single-use transaction.
```

Comment:
can you put the warning here and in insertBatch,delete about this not being replay safe similar to how we have in java:
https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java#L43